cvat/apps/engine/media_extractors.py

```diff
@@ -155,96 +154,88 @@ def __init__(self, source_path, step=1, start=0, stop=0):
-            start=0,
-            stop=0,
+            step=step,
+            start=start,
+            stop=stop,
         )
 
     def __del__(self):
-        if (self._tmp_dir):
-            self.delete_tmp_dir(self._tmp_dir)
+        delete_tmp_dir(self._tmp_dir)
 
-    @property
-    def image_names(self):
-        return  [os.path.join(os.path.dirname(self._pdf_source), os.path.relpath(p, self._tmp_dir)) for p in super().image_names]
+    def get_path(self, i):
+        base_dir = os.path.dirname(self._pdf_source)
+        return os.path.join(base_dir, os.path.relpath(self._source_path[i], self._tmp_dir))
 
-class ZipReader(IMediaReader):
+class ZipReader(ImageListReader):
     def __init__(self, source_path, step=1, start=0, stop=0):
         self._zip_source = zipfile.ZipFile(source_path[0], mode='r')
         file_list = [f for f in self._zip_source.namelist() if get_mime(f) == 'image']
         super().__init__(file_list, step, start, stop)
 
-    def __iter__(self):
-        for f in zip(self._source_path, self.image_names):
-            yield (BytesIO(self._zip_source.read(f[0])), f[1])
-
-    def __len__(self):
-        return len(self._source_path)
-
-    def __getitem__(self, k):
-        return (BytesIO(self._zip_source.read(self._source_path[k])), self.image_names[k])
-
     def __del__(self):
         self._zip_source.close()
 
-    def save_preview(self, preview_path):
-        with open(preview_path, 'wb') as f:
-            f.write(self._zip_source.read(self._source_path[0]))
+    def get_preview(self):
+        io_image = io.BytesIO(self._zip_source.read(self._source_path[0]))
+        return self._get_preview(io_image)
 
     def get_image_size(self):
-        img = Image.open(BytesIO(self._zip_source.read(self._source_path[0])))
+        img = Image.open(io.BytesIO(self._zip_source.read(self._source_path[0])))
         return img.width, img.height
 
-    @property
-    def image_names(self):
-        return [os.path.join(os.path.dirname(self._zip_source.filename), p) for p in self._source_path]
+    def get_image(self, i):
+        return io.BytesIO(self._zip_source.read(self._source_path[i]))
+
+    def get_path(self, i):
+        return os.path.join(os.path.dirname(self._zip_source.filename), self._source_path[i])
 
 class VideoReader(IMediaReader):
     def __init__(self, source_path, step=1, start=0, stop=0):
-        self._output_fps = 25
-
         super().__init__(
             source_path=source_path,
             step=step,
             start=start,
             stop=stop,
         )
 
-    def __iter__(self):
-        def decode_frames(container):
-            for packet in container.demux():
-                if packet.stream.type == 'video':
-                    for frame in packet.decode():
-                        yield frame
+    def _has_frame(self, i):
+        if i >= self._start:
+            if (i - self._start) % self._step == 0:
+                if self._stop == 0 or i <= self._stop:
```

Comment:
@azhavoro , it seems that for images and for videos self._stop has different meaning (in case of images it doesn't include the latest frame). Can we easily unitfy?

Reply:
already unified, we add +1 to the stop frame value in case of images.
https://github.com/opencv/cvat/blob/small_preview_and_progress/cvat/apps/engine/media_extractors.py#L75

Reply:
@azhavoro , why in this case we have `i <= self._stop` here?

Reply:
VideoReader class is not inherited from ImageListReader, stop + 1 doesn't affect VideoReader
stop value and iteration for ImageListReader:
```
...
stop = min(len(source_path), stop + 1)
...
for i in range(self._start, self._stop, self._step):
...
```

stop value and iteration for VideoReader:
```
...
stop=stop,
...
if i >= self._start:
    if (i - self._start) % self._step == 0:
        if self._stop == 0 or i <= self._stop:

```
Have I missed something?

Reply:
@azhavoro , let me repeat the question: 

> @azhavoro , it seems that for images and for videos self._stop has different meaning (in case of images it doesn't include the latest frame). Can we easily unify?

Reply:
@azhavoro , I just want to confirm that for videos we can use `i < self._stop` (or if we cannot let's change logic). In the future it will be very difficult to read the code and recall that two similar classes have different logic about the same variable.

Reply:
@nmanovic I unified the iteration logic, also fixed corner case when user specified stop frame = 0.